src/mlpack/tests/bayesian_linear_regression_test.cpp

```diff
@@ -0,0 +1,156 @@
+/**
+ * @file bayesian_linear_regression_test.cpp
+ * @author Clement Mercier
+ *
+ * Test for BayesianLinearRegression.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include <mlpack/core/data/load.hpp>
+#include <mlpack/methods/bayesian_linear_regression/bayesian_linear_regression.hpp>
+#include <mlpack/methods/linear_regression/linear_regression.hpp>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace mlpack::regression;
+using namespace mlpack::data;
+
+BOOST_AUTO_TEST_SUITE(BayesianLinearRegressionTest);
+
+void GenerateProblem(arma::mat& X,
+                     arma::rowvec& y,
+                     size_t nPoints,
+                     size_t nDims,
+                     float sigma = 0.0)
+{
+  X = arma::randn(nDims, nPoints);
+  arma::colvec omega = arma::randn(nDims);
+  // Compute y and add noise.
+  y = omega.t() * X + arma::randn(nPoints).t() * sigma;
+}
+
+// Ensure that predictions are close enough to the target
+// for a free noise dataset.
+BOOST_AUTO_TEST_CASE(BayesianLinearRegressionRegressionTest)
+{
+  arma::mat X;
+  arma::rowvec y, predictions;
+
+  GenerateProblem(X, y, 200, 10);
+
+  // Instanciate and train the estimator.
+  BayesianLinearRegression estimator(true);
+  estimator.Train(X, y);
+  estimator.Predict(X, predictions);
+
+  // Check the predictions are close enough to the targets in a free noise case.
+  for (size_t i = 0; i < y.size(); i++)
+    BOOST_REQUIRE_CLOSE(predictions[i], y[i], 1e-6);
+
+  // Check that the estimated variance is zero.
+  BOOST_REQUIRE_SMALL(estimator.Variance(), 1e-6);
+}
+
+// Verify fitIntercept and normalize equal false do not affect the solution.
+BOOST_AUTO_TEST_CASE(TestCenter0Normalize0)
+{
+  arma::mat X;
+  arma::rowvec y;
+  size_t nDims = 30, nPoints = 100;
+
+  GenerateProblem(X, y, nPoints, nDims, 0.5);
+
+  BayesianLinearRegression estimator(false, false);
+
+  estimator.Train(X, y);
+
+  // To be neutral dataOffset must be all 0.
+  BOOST_REQUIRE(sum(estimator.DataOffset()) == 0.0);
```

Comment:
An additional optimization might just be to ensure that `DataOffset()` is not resized to `nDims` when `center` is `false`, and `DataScale()` is not resized to `nDims` when `normalize` is `false`.  That will also save some space when serializing. :+1:

Reply:
Okay, I will change this point. The CenterScaleData() will certainly disappear and the Predict() methods and some tests will change.

Reply:
Do you want to allow a scaling without a centering ?

Reply:
I see what you mean.  I think that actually you solved the problem really nicely with the `CenterScaleData()` and `CenterScaleDataPred()` functions---but maybe did I overlook something?  It seems like now you could keep `dataOffset` and `dataScale` empty when `center` and `normalize` are `false`, respectively.